### ISSUE TITLE
Incremental type narrowing, enabling gradual object initialization

### ISSUE THREAD

**Issue by Richiban** [microsoft/TypeScript#35086]
## Search Terms

delayed initialization, gradual initialization

## Suggestion

What if, on every assignment to either a value or a property, that the type of that value is narrowed? It's probably easiest explained with an example:

## Use Cases

It means that this code for gradually initializing an object would just work:

```typescript
type Person = { firstName: string, surname: string }

function f() : Person {
    const p : Partial<Person> = {}

    p.firstName = ""
    p.surname = ""

    return p;
}
```

Currently this code does not type check because the compiler doesn't change the type of `p` after the assignments to its properties.

My thinking is that, for any value `x` of type `X`, 

```typescript
x.a = "some string"
```

would result in `x` having its type narrowed to:

```typescript
X & { a : string }
```

This means that the partial initialization example would work. If it works for variables as well as properties then so would this:

```typescript
function f(mode: "read" | "read/write") {
    // ... Code ...
}

let s: string;

s = "read"

f(s)
```

## Checklist

My suggestion meets these guidelines:

* [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [X] This wouldn't change the runtime behavior of existing JavaScript code
* [X] This could be implemented without emitting different JS based on the types of the expressions
* [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [X] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by AnyhowStep** [comment]
This is kind of already supported with functions.
Maybe not too much of a stretch to support it for objects?

**Comment by nmain** [comment]
The examples given seem sensible enough, but it seems like the proposal as written would allow all sorts of floop through:

```ts
declare var a: { prop: string };
a.brop = "bar"; // woops!  meant to write `prop`.
```

This example would give no error because it would just evolve the type of `a`?

**Comment by Richiban** [comment]
@nmain No, it would only allow for type _narrowing_, not completely changing the type as in your example.

**Comment by nmain** [comment]
Your change makes it clearer.  Evolving type `{ prop: string }` to `{ prop: string, brop: string }` is a narrowing, of course, but I like the idea that it wouldn't work unless `{ brop: unknown }` was already part of the type

**Comment by Richiban** [comment]
@nmain Correct!

To put it another way for other readers, given 

```typescript
declare var a: { prop: string };
```

the statement

```typescript
a.brop = "bar";
```

is still illegal; this proposal doesn't change that. What happens is that, after an assignment statement, the type of the value on the LHS updates to a more specific type depending on what was on the RHS of the assignment.

**Comment by jack-williams** [comment]
Is this basically (narrow object by property) #33205 + narrow on assignment?

**Comment by Richiban** [comment]
@jack-williams It's pretty close, differing in that that issue deals with narrowing object properties in general; I'm talking about narrowing specifically on assignment, and for both properties and variables.

**Comment by Richiban** [comment]
I would like to add some more examples to this, since this is still a feature I would very much appreciate. 

A common pattern in our codebase is to take an `option` object as the last parameter of a function, all of whose properties are optional. We then give any missing properties defaults by writing out an object literal with the defaults and spreading the provided `options` object in the same literal. Here's an example:

```ts
export function* iterBlockStarts(
  fromPosition: Point,
  options: {
    direction?: Direction;
    indentationLevel?: IndentationRequest;
    restrictToCurrentScope?: boolean;
  }
) {
  const finalOptions = {
    direction: "forwards",
    indentationLevel: "same-indentation",
    restrictToCurrentScope: false,
    ...options,
  };

  ....
}
```

It's not a big deal but it's irritating having to create a new constant `finalOptions`; something that you wouldn't have to do if assignment narrowed a variable's type. You'd be able to simply reassign `options`.
